### PR TITLE
Java user propagator interface

### PR DESCRIPTION
**Pull request by d1tto**
https://github.com/Z3Prover/z3/issues/6097

### CODE REVIEW

**Comment by d1tto** [comment]
@microsoft-github-policy-service agree

File: scripts/update_api.py [hunk 1]
```diff
@@ -633,7 +633,71 @@ def mk_java(java_src, java_dir, package_name):
...
+        solver = _solver;
+        javainfo = propagateInit(this, ctx, solver);
+    }
+
+    public final void close() {
+        Native.propagateDestroy(this, ctx, solver, javainfo);
+        javainfo = 0;
+        solver = 0;
+        ctx = 0;
+    }
```
Comment (ThomasHaas) [inline scripts/update_api.py:664]: You can implement the `AutoClosable` interface to more clearly communicate that this object maintains resources that need to get released.

File: src/api/java/NativeStatic.txt [hunk 2]
```diff
@@ -77,3 +77,153 @@ DLL_VIS JNIEXPORT void JNICALL Java_com_microsoft_z3_Native_setInternalErrorHand
...
+  info->jenv = jenv;
+  info->jobj = jenv->NewGlobalRef(jobj);
+  jclass jcls = jenv->GetObjectClass(info->jobj);
+  info->push = jenv->GetMethodID(jcls, "pushWrapper", "()V");
+  info->pop = jenv->GetMethodID(jcls, "popWrapper", "(I)V");
+  info->fresh = jenv->GetMethodID(jcls, "freshWrapper", "(J)Lcom/microsoft/z3/Native$UserPropagatorBase;");
+  info->created = jenv->GetMethodID(jcls, "createdWrapper", "(J)V");
+  info->fixed = jenv->GetMethodID(jcls, "fixedWrapper", "(JJ)V");
+  info->eq = jenv->GetMethodID(jcls, "eqWrapper", "(JJ)V");
+  info->final = jenv->GetMethodID(jcls, "finWrapper", "()V");
```
Comment (ThomasHaas) [inline src/api/java/NativeStatic.txt:171]: I think you should first initialize the `info` object before calling `Z3_solver_propagate_init`.
Also, this part requires error-handling.
All the `jenv->GetMethodID` calls are able to return a null value if they fail for some reason (e.g., if the method name is wrong).
I'm not familar with exception handling in JNI code, but the JNI provide a lot of exception-related functionality.

File: src/api/java/NativeStatic.txt [hunk 3]
```diff
@@ -77,3 +77,153 @@ DLL_VIS JNIEXPORT void JNICALL Java_com_microsoft_z3_Native_setInternalErrorHand
...
+  JNIEnv *jenv;
+  jclass jcls;
+  jobject jobj;
+
+  jmethodID push;
+  jmethodID pop;
+  jmethodID fresh;
+  jmethodID created;
+  jmethodID fixed;
+  jmethodID eq;
```
Comment (NikolajBjorner) [inline src/api/java/NativeStatic.txt:93]: Initialize these fields? 
jmethodID eq = nullptr; ?

**Comment by NikolajBjorner** [comment]
Can you add a different pull request with the unit test added to examples/java?

**Comment by NikolajBjorner** [comment]
I am merging this as is: if it is broken it is new functionality and does not affect previous functionality.

**Comment by d1tto** [comment]
Thanks. I'll add example soon.

**Comment by NikolajBjorner** [comment]
note also build warnings

src\api\java\Native.cpp(211): warning C4244: 'argument': conversion from 'jlong' to 'unsigned int', possible loss of data
src\api\java\Native.cpp(211): warning C4244: 'argument': conversion from 'jlong' to 'unsigned int', possible loss of data
src\api\java\Native.cpp(232): warning C4312: 'type cast': conversion from 'long' to 'Z3_ast' of greater size

**Pull request merged**